File: app/models/donation.py
from datetime import datetime
from sqlalchemy import Numeric
from app import db

class Donation(db.Model):
    __tablename__ = 'donations'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    
    # Montant et devise
    amount = db.Column(Numeric(10, 2), nullable=False)
    currency = db.Column(db.String(3), default='CAD')
    
    # Type de don
    donation_type = db.Column(db.String(50), default='Dîme')  # 'Dîme', 'Offrande', 'Mission', 'Construction', 'Autre'
    purpose = db.Column(db.String(200))  # Description spécifique du don
    
    # Informations de paiement
    payment_method = db.Column(db.String(50))  # 'Stripe', 'PayPal', 'Cash', 'Check'
    payment_id = db.Column(db.String(100))  # ID de transaction externe
    payment_status = db.Column(db.String(20), default='Pending')  # 'Pending', 'Completed', 'Failed', 'Refunded'
    
    # Reçu fiscal
    receipt_number = db.Column(db.String(50), unique=True)
    is_tax_deductible = db.Column(db.Boolean, default=True)
    receipt_sent = db.Column(db.Boolean, default=False)
    receipt_sent_at = db.Column(db.DateTime)
    
    # Dates
    donation_date = db.Column(db.DateTime, default=datetime.utcnow)
    processed_at = db.Column(db.DateTime)
    
    # Métadonnées
    notes = db.Column(db.Text)
    is_anonymous = db.Column(db.Boolean, default=False)
    
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relations
    user = db.relationship('User', back_populates='donations')
    
    def generate_receipt_number(self):
        """Génère un numéro de reçu unique"""
        if not self.donation_date:
            self.donation_date = datetime.utcnow()
        year = self.donation_date.year
        # Format: ICCS-YYYY-XXXXXX (Impact Centre Chrétien Sherbrooke)
        count = Donation.query.filter(
            db.extract('year', Donation.donation_date) == year
        ).count() + 1
        self.receipt_number = f"ICCS-{year}-{count:06d}"
    
    @property
    def formatted_amount(self):
        return f"{self.amount:.2f} {self.currency}"
    
    def __repr__(self):
        return f'<Donation {self.formatted_amount} by {self.user.full_name}>'

File: impact-centre-mobile/backend/models/Donation.js
import mongoose from 'mongoose';

const donationSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  amount: {
    type: Number,
    required: true,
    min: 0
  },
  currency: {
    type: String,
    default: 'CAD',
    enum: ['CAD', 'USD', 'EUR']
  },
  donationType: {
    type: String,
    enum: ['Dîme', 'Offrande', 'Mission', 'Construction', 'Autre'],
    default: 'Dîme'
  },
  purpose: {
    type: String,
    trim: true
  },
  paymentMethod: {
    type: String,
    enum: ['Stripe', 'PayPal', 'Cash', 'Check', 'Transfer'],
    default: 'Stripe'
  },
  paymentId: {
    type: String,
    trim: true
  },
  paymentStatus: {
    type: String,
    enum: ['Pending', 'Completed', 'Failed', 'Refunded'],
    default: 'Pending'
  },
  receiptNumber: {
    type: String,
    unique: true,
    sparse: true
  },
  isTaxDeductible: {
    type: Boolean,
    default: true
  },
  receiptSent: {
    type: Boolean,
    default: false
  },
  receiptSentAt: {
    type: Date
  },
  donationDate: {
    type: Date,
    default: Date.now
  },
  processedAt: {
    type: Date
  },
  notes: {
    type: String,
    trim: true
  },
  isAnonymous: {
    type: Boolean,
    default: false
  },
  stripeData: {
    paymentIntentId: String,
    chargeId: String,
    customerId: String
  }
}, {
  timestamps: true
});

// Index pour recherche et tri
donationSchema.index({ userId: 1, donationDate: -1 });
donationSchema.index({ donationDate: -1 });
donationSchema.index({ paymentStatus: 1 });
donationSchema.index({ receiptNumber: 1 });

// Virtual pour le montant formaté
donationSchema.virtual('formattedAmount').get(function() {
  return `${this.amount.toFixed(2)} ${this.currency}`;
});

// Méthode pour générer le numéro de reçu
donationSchema.methods.generateReceiptNumber = async function() {
  if (this.receiptNumber) return this.receiptNumber;
  
  const year = this.donationDate.getFullYear();
  const count = await this.constructor.countDocuments({
    donationDate: {
      $gte: new Date(year, 0, 1),
      $lt: new Date(year + 1, 0, 1)
    },
    receiptNumber: { $exists: true }
  });
  
  this.receiptNumber = `ICCS-${year}-${String(count + 1).padStart(6, '0')}`;
  return this.receiptNumber;
};

// Méthode pour marquer comme traité
donationSchema.methods.markAsProcessed = function() {
  this.paymentStatus = 'Completed';
  this.processedAt = new Date();
  return this;
};

// Méthode statique pour les statistiques
donationSchema.statics.getYearlyStats = function(year, userId = null) {
  const match = {
    donationDate: {
      $gte: new Date(year, 0, 1),
      $lt: new Date(year + 1, 0, 1)
    },
    paymentStatus: 'Completed'
  };
  
  if (userId) {
    match.userId = new mongoose.Types.ObjectId(userId);
  }
  
  return this.aggregate([
    { $match: match },
    {
      $group: {
        _id: '$donationType',
        totalAmount: { $sum: '$amount' },
        count: { $sum: 1 }
      }
    },
    { $sort: { totalAmount: -1 } }
  ]);
};

export default mongoose.models.Donation || mongoose.model('Donation', donationSchema);